core-rust/state-manager/src/staging/epoch_handling.rs

```diff
@@ -62,47 +62,135 @@
+            store,
+            self.epoch_state_version,
+            previous_epoch_root,
+            self.current_accu_tree_len(),
+        )
+    }
+
+    /// Creates an accu tree merger which can merge the next slice(s) of the scoped epoch's accu
+    /// tree at the current version (i.e. declared during this factory's construction).
+    pub fn create_merger<N>(&self) -> AccuTreeSliceMerger<N> {
+        AccuTreeSliceMerger::new(self.current_accu_tree_len())
+    }
+
     /// Returns the actual number of leaves in the epoch's accu tree.
     /// This takes into account the extra first leaf (i.e. previous epoch's tree root), and handles
     /// the "fresh epoch" (i.e. empty accu tree) case.
-    pub fn current_accu_tree_len(&self) -> usize {
+    fn current_accu_tree_len(&self) -> usize {
         if self.epoch_version_count == 0 {
             0
         } else {
             self.epoch_version_count + 1
         }
     }
+}
 
-    /// Adjusts the next planned batch of leaves to be appended to the epoch's accu tree.
-    /// The adjustment consists of prepending the previous epoch's root in case the current epoch
-    /// has just been started.
-    pub fn adjust_next_batch<T>(&self, previous_epoch_root: T, mut next_batch: Vec<T>) -> Vec<T> {
-        if self.epoch_version_count == 0 {
-            next_batch.insert(0, previous_epoch_root);
+/// An [`AccuTree`] wrapper which adds the proper "accu tree per epoch" handling.
+pub struct EpochAccuTreeBuilder<'s, S, N> {
+    previous_epoch_root: Option<N>,
+    epoch_tree_len: usize,
+    epoch_scoped_store: EpochScopedAccuTreeStore<'s, S>,
+}
+
+impl<'s, S: AccuTreeStore<StateVersion, N>, N: IsMerklizableHash> EpochAccuTreeBuilder<'s, S, N> {
+    fn new(
+        store: &'s mut S,
+        epoch_state_version: StateVersion,
+        previous_epoch_root: N,
+        epoch_tree_len: usize,
+    ) -> Self {
+        Self {
+            previous_epoch_root: Some(previous_epoch_root),
+            epoch_tree_len,
+            epoch_scoped_store: EpochScopedAccuTreeStore {
+                store,
+                epoch_state_version,
+            },
+        }
+    }
+
+    /// Appends the next leaf to the epoch's accu tree.
+    /// This method will handle a special case of an epoch that was just started (which requires
+    /// inserting the previous epoch's root as the first leaf of this epoch's tree).
+    pub fn append(&mut self, new_leaf_hash: N) {
+        self.append_batch(vec![new_leaf_hash]);
+    }
+
+    /// A batch variant of the [`append()`] method.
+    pub fn append_batch(&mut self, mut new_leaf_hashes: Vec<N>) {
+        if self.epoch_tree_len == 0 {
+            new_leaf_hashes.insert(0, self.previous_epoch_root.take().unwrap());
         }
-        next_batch
+        let appended_len = new_leaf_hashes.len();
+        AccuTree::new(&mut self.epoch_scoped_store, self.epoch_tree_len).append(new_leaf_hashes);
+        self.epoch_tree_len += appended_len;
+    }
+}
+
```

Comment:
_(note from author)_

I moved some code from `result.rs` to here;
Regarding the new code: I only formalized the raw "extracted logic" (i.e. `AccuTreeEpochHandler`) into a proper "epoch aware accu tree factory (+wrapper)".